### ISSUE TITLE
[Bug]: [benchmark][natsmq] Concurrency test after inserting 100 million data, milvus restarted 4 times with natsmq 

### ISSUE THREAD

**Issue by elstic** [milvus-io/milvus#26421]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:master-20230816-f9341f66
- Deployment mode(standalone or cluster):standalone
- MQ type(rocksmq, pulsar or kafka):    natsmq
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

deploy_config:fouramf-server-natsmq
other_params: test_concurrent_locust_100m_ivf_sq8_ddl_dql_standalone 

```
    extraConfigFiles:
      user.yaml: |+
        mq:
          type: natsmq
    pulsar:
      enabled: false
```

server
Problem: milvus rebooted 4 times 
```
fouramf-5n85g-89-4351-etcd-0                                      1/1     Running                  0               13h     10.104.20.44    4am-node22   <none>           <none>
fouramf-5n85g-89-4351-milvus-standalone-64c477449c-dfkdp          1/1     Running                  4 (133m ago)    13h     10.104.18.206   4am-node25   <none>           <none>
fouramf-5n85g-89-4351-minio-84c9d7b46d-rx26j                      1/1     Running                  0               13h     10.104.18.205   4am-node25   <none>           <none> 
```

memory : 
The monitor also observed a reboot of the milvus
![image](https://github.com/milvus-io/milvus/assets/48523564/33dfc8de-ee0c-4560-951c-0ec49aeba2b1)

</b></b></b>
There is no search failure on the client side, only a reboot of milvus.

### Expected Behavior

No milvus reboot during the entire test period

### Steps To Reproduce

```markdown
1. create a collection or use an existing collection  
  2. build an IVF_SQ8 index on the vector column
  3. insert 100 million vectors
  4. flush collection
  5. build index on vector column with the same parameters  
  6. count the total number of rows
  7. load collection
  8. execute concurrent search, query,load,scene_test 
    (scene_test steps: 
       1） Create a collection 2） Insert 3000 pieces of data  3） flush collection 
       4） Create an index  5） drop collection)
  9. clean all collections
```

**Comment by jaime0815** [comment]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x172a3e2]

**Comment by chyezh** [comment]
related issue of go, same panic:
https://github.com/golang/go/issues/58846

**Comment by chyezh** [comment]
reproduced on fouramf-5n85g-2-82-1426-milvus-standalone-5697d4ffcc-qpnlh

**Comment by chyezh** [comment]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x172a3e2]
goroutine 26989239 [running]:
panic({0x3a7e2a0, 0x5abe2a0})
	/usr/local/go/src/runtime/panic.go:941 +0x397 fp=0xc0063529c0 sp=0xc006352900 pc=0x153b377
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:220
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:818 +0x336 fp=0xc006352a10 sp=0xc0063529c0 pc=0x1552416
context.(*valueCtx).Value(0xc00a3eb470?, {0x392a7c0?, 0x43e3a28?})
	/usr/local/go/src/context/context.go:563 +0x22 fp=0xc006352a48 sp=0xc006352a10 pc=0x172a3e2
go.opentelemetry.io/otel/trace.SpanFromContext({0x4416378?, 0x0?})
	/go/pkg/mod/go.opentelemetry.io/otel/trace@v1.13.0/context.go:52 +0x3e fp=0xc006352a80 sp=0xc006352a48 pc=0x1d4473e
go.opentelemetry.io/otel/sdk/trace.(*tracer).Start(0xc00280b4c0, {0x4416378?, 0x0?}, {0x3eeec00, 0x7}, {0x0?, 0xc006352ce0?, 0xc006352ce8?})
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.13.0/trace/tracer.go:46 +0x105 fp=0xc006352c78 sp=0xc006352a80 pc=0x1fc5525
github.com/milvus-io/milvus/internal/util/flowgraph.(*InputNode).Operate(0xc00322ec40, {0xc00e7336e0?, 0x60?, 0x40177f0?})
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/input_node.go:119 +0xe04 fp=0xc006352e90 sp=0xc006352c78 pc=0x2e17fa4
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc0045c1450)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:127 +0x2ac fp=0xc006352fc8 sp=0xc006352e90 pc=0x2e188cc
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start.func1()
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:71 +0x26 fp=0xc006352fe0 sp=0xc006352fc8 pc=0x2e18306
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1571 +0x1 fp=0xc006352fe8 sp=0xc006352fe0 pc=0x15709e1
created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:71 +0x79

**Comment by chyezh** [comment]
This doesn't seem like a null pointer problem, panic program access the 0x10 address, but not 0x0 address.

**Comment by chyezh** [comment]
> This doesn't seem like a null pointer problem, panic program access the 0x10 address, but not 0x0 address.

It's just the memory layout of go, *valueCtx is a nil but Context is not, reproduced by those code:

- a is the valueCtx;
- A is the Context;

```
type A interface {
	Print()
}

type a struct {
	A
	p int
}

func (aa *a) Print() {
	aa.p++
}

func main() {
	var l A = (*a)(nil)
	l.Print()
}

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x104cfdd00]

goroutine 1 [running]:
main.(*a).Print(0x104d1c160?)
        /Users/zilliz/repo/github/chyezh/test-tools/temp/main.go:13
main.main()
        /Users/zilliz/repo/github/chyezh/test-tools/temp/main.go:18 +0x20
exit status 2
```

**Comment by chyezh** [comment]
Found that msgdispatcher will share non-dml msg instance, the underlying context.Context will be set in multiple concurrent unit. That maybe cause the dirty read to *valueCtx pointer when data race.

But it's not reproduced by unittest, I will fix this bug and try these test again.

![image](https://github.com/milvus-io/milvus/assets/20332315/03f12c14-3441-45ab-a32b-9efbeb69732b)

**Comment by chyezh** [comment]
new image with PR do not reproduce this bug.

fouramf-5n85g-5

**Comment by elstic** [comment]
This issue has been fixed, verify mirror: master-20230822-fe16cdbc
